### PR TITLE
team-level std algos: part 8

### PR DESCRIPTION
**Pull request by fnrizzi**
Co-authored-by: Cezary Skrzyński
Co-authored-by: Jakub Strzebonski

## Content 

Team-level implementation and tests for:

- `Kokkos_Copy.hpp` 
- `Kokkos_CopyN.hpp`
- `Kokkos_CopyBackward.hpp`
- `Kokkos_CopyIf.hpp` 
- `Kokkos_UniqueCopy.hpp`
- `Kokkos_Remove.hpp` 
- `Kokkos_RemoveIf.hpp`
- `Kokkos_RemoveCopy.hpp`
- `Kokkos_RemoveCopyIf.hpp`

## Technical comments

-  this PR is not split any further because  of the following dependencies: 
   - `Kokkos_UniqueCopy.hpp`: needs `copy`
   - `Kokkos_Remove.hpp`: needs `copy_if`
   - `Kokkos_RemoveIf.hpp`: needs `copy_if` and `count_if`
   - `Kokkos_RemoveCopy.hpp`: needs `copyif`
   - `Kokkos_RemoveCopyIf.hpp`: needs `copyif`

- the change `class` -> `typename` and the enable if defaulted in the template parameters comes from [this request](https://github.com/kokkos/kokkos/pull/6200#issuecomment-1626406816)

- the code duplication is intentional and agreed upon after the long discussion in [here](https://github.com/kokkos/kokkos/pull/6200)

- tests are run for both contiguous views and strided views

## Other comments

- all team-level PRs are tracked in: https://github.com/kokkos/kokkos/issues/6352

- this comes from breaking apart https://github.com/kokkos/kokkos/pull/5150

## Possible conflicts

Implementation-wise, this has not conflict with other team-level PRs so it could be merged independently. 
However, a trivial conflict in the CMakeLists inside the unit test occurs if another of the team-level PRs is merged before this because each team-level PR adds test using a specific "label". Resolving this is trivial. 

## Associated wiki PRs

- `Kokkos_Copy.hpp`,  `Kokkos_CopyIf.hpp`, `Kokkos_CopyN.hpp`, `Kokkos_CopyBackward.hpp`: see [wiki PR](https://github.com/kokkos/kokkos-core-wiki/pull/386)
- `Kokkos_UniqueCopy.hpp`, `Kokkos_Remove.hpp`, `Kokkos_RemoveIf.hpp`,  `Kokkos_RemoveCopy.hpp`: see [wiki PR](https://github.com/kokkos/kokkos-core-wiki/pull/434)
- `Kokkos_RemoveCopyIf.hpp`: see [wiki PR](https://github.com/kokkos/kokkos-core-wiki/pull/436)

IMPORTANT: wiki PRs are not complete yet but 99% there 

The issue tracking all *wiki PRs* is [here](https://github.com/kokkos/kokkos-core-wiki/issues/379)

### CODE REVIEW

**Review by Rombur — APPROVED** [review]

File: algorithms/src/std_algorithms/impl/Kokkos_UniqueCopy.hpp [hunk 1]
```diff
@@ -78,38 +77,105 @@ OutputIterator unique_copy_impl(const std::string& label,
...
+      }
+
+      // we need to copy the last element always
+      d_first[countView()++] = first[num_elements - 1];
+    });
+    teamHandle.team_barrier();
+
+    // return the correct iterator: we need +1 here because we need to
+    // return iterator to the element past the last element copied
+    return d_first + count;
```
Comment (crtrott) [inline algorithms/src/std_algorithms/impl/Kokkos_UniqueCopy.hpp:166]: this is definitely wrong. Only thread 0 will get back a count != 0. You need to use a broadcast single. Here, we also should improve the test figuring out why this doesn't fail the test.

**Review by crtrott — CHANGES_REQUESTED** [review]

File: algorithms/unit_tests/TestStdAlgorithmsTeamUniqueCopy.cpp [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+
+    if (m_apiPick == 0) {
+      auto it =
+          KE::unique_copy(member, KE::begin(myRowViewFrom),
+                          KE::end(myRowViewFrom), KE::begin(myRowViewDest));
+
+      Kokkos::single(Kokkos::PerTeam(member), [=, *this]() {
+        m_distancesView(myRowIndex) =
+            KE::distance(KE::begin(myRowViewDest), it);
+      });
```
Comment (crtrott) [inline algorithms/unit_tests/TestStdAlgorithmsTeamUniqueCopy.cpp:60]: this test doesn't trigger if any thread other than thread 0 will give the wrong answer.

File: algorithms/src/std_algorithms/impl/Kokkos_UniqueCopy.hpp [hunk 3]
```diff
@@ -78,38 +77,105 @@ OutputIterator unique_copy_impl(const std::string& label,
...
+        const auto& val_i   = first[i];
+        const auto& val_ip1 = first[i + 1];
+        if (!pred(val_i, val_ip1)) {
+          d_first[countView()++] = val_i;
+        }
+      }
+
+      // we need to copy the last element always
+      d_first[countView()++] = first[num_elements - 1];
+    });
```
Comment (crtrott) [inline algorithms/src/std_algorithms/impl/Kokkos_UniqueCopy.hpp:161]:
```suggestion
    std::size_t count = 0;
    Kokkos::single(Kokkos::PerTeam(teamHandle), [=](std::size_t& lcount) {
      lcount = 0;
      for (std::size_t i = 0; i < num_elements - 1; ++i) {
        const auto& val_i   = first[i];
        const auto& val_ip1 = first[i + 1];
        if (!pred(val_i, val_ip1)) {
          d_first[lcount++] = val_i;
        }
      }

      // we need to copy the last element always
      d_first[lcount++] = first[num_elements - 1];
    }, count);
```

File: algorithms/unit_tests/TestStdAlgorithmsTeamUniqueCopy.cpp (hunk 2 above)
Comment (fnrizzi) [reply algorithms/unit_tests/TestStdAlgorithmsTeamUniqueCopy.cpp:64]: improved tests for this , let me know if the new solutions works for you

**Review by crtrott — APPROVED** [review]

**Pull request merged**